File: etc/BaseX.php
<?php
/*
 * This PHP module provides all methods to connect and communicate with the
 * BaseX Server.
 *
 * The session class manages the communication between server and client.
 * This class has to be called by your client code (see Example.php).
 *
 * The Constructor of the Session class expects a hostname, port, username and
 * password for the connection. The socket connection will then be established
 * via the hostname and the port.
 *
 * For the execution of commands you need to call the execute method with the command
 * as argument. The result and the info will then be written to the corresponding string.
 * These strings can be fetched with the methods result() and info().
 *
 * Example:
 * 
 * include("BaseX.php");
 * 
 * try {
 * 	// create session
 *	$session = new Session("localhost", 1984, "admin", "admin");
 *	// perform command; show info if something went wrong
 * 	if(!$session->execute("xquery 1 + 2")) {
 *   	print $session->info();
 * 	} else {
 *   	print $session->result();
 *	}
 *	// close session
 *	$session->close();
 *	} catch (Exception $e) {
 *	  	print $e->getMessage();
 *	}
 *
 * (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
 */
class Session {
  /* Class variables.*/
  var $socket, $result, $info, $buffer, $bpos, $bsize;

  /* Constructor. */
  function __construct($h, $p, $user, $pw) {
    // create server connection
    $this->socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
    if(!socket_connect($this->socket, $h, $p)) {
      throw new Exception("Can't communicate with server.");
    }

    // receive timestamp
    $ts = $this->readString();

    // send username and hashed password/timestamp
    $md5 = hash("md5", hash("md5", $pw).$ts);
    socket_write($this->socket, "$user\0$md5\0");

    // receives success flag
    if(socket_read($this->socket, 1) != "\0") {
      throw new Exception("Access denied.");
    }
  }

  /* Executes a command and writes the result and the info to the
   * corresponding strings. */
  public function execute($com) {
    // send command to server
    socket_write($this->socket, "$com\0");

    // receive result
    $this->init();
    $this->result = $this->readString();
    $this->info = $this->readString();
    
    return $this->read() == "\0";
  }

  /* Returns the result. */
  public function result() {
    return $this->result;
  }

  /* Returns the info string. */
  public function info() {
    return $this->info;
  }

  /* Closes the socket. */
  public function close() {
    socket_write($this->socket, "exit\0");
    socket_close($this->socket);
  }

  /* Initiates the incoming message. */
  private function init() {
    $this->bpos = 0;
    $this->bsize = 0;
  }

  /* Receives a string from the socket. */
  private function readString() {
    $com = "";
    while(($d = $this->read()) != "\0") {
      $com .= $d;
    }
    return $com;
  }

  /* Returns the next byte. */
  private function read() {
    if($this->bpos == $this->bsize) {
      $this->bsize = socket_recv($this->socket, $this->buffer, 4096, 0);
      $this->bpos = 0;
    }
    return $this->buffer[$this->bpos++];
  }
}
?>

File: etc/BaseX.py
# This Python module all methods to connect and communicate with the
# BaseX Server.
#
# The Session class manages the communication between server and client.
# This class has to be called by your client code (see Example.py).
#
# The Constructor of the Session class expects a hostname, port, username and
# password for the connection. The socket connection will then be established
# via the hostname and the port.
#
# The Client class is a console client which allows to interactively input
# commands.
#
# For the execution of commands you need to call the execute method with the command
# as argument. The result and the info will then be written to the corresponding string.
# These strings can be fetched with the methods result() and info().
#
# Example:
# 
# import BaseX
# try:
# # create session
# cs = BaseX.Session('localhost', 1984, 'admin', 'admin')
# # perform command; show info if something went wrong
# if not cs.execute("xquery 1 + 2):
#    print cs.info()
#  else:
#    print cs.result()
# # close session
# cs.close()
# except IOError as e:
#  print e
#
# (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License

import hashlib, socket, array, getopt, sys, getpass

class Session():

  # Constructor.
  def __init__(self, host, port, user, pw):
    # allocate 4kb buffer
    self.__buf = array.array('B', '\0' * 0x1000)
    self.__init()

    # create server connection
    global s
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, port))

    # receive timestamp
    ts = self.__readString()
    
    # code password and timestamp in md5
    # send user name and hashed password/timestamp
    m = hashlib.md5()
    m.update(hashlib.md5(pw).hexdigest())
    m.update(ts)

    s.send(user + '\0')
    s.send(m.hexdigest() + '\0')

    # receives success flag
    if s.recv(1) != '\0':
      raise IOError('Access Denied.')

  # Executes a command.
  def execute(self, com):
    # send command to server
    s.send(com + '\0')

    # receive result
    self.__init()
    self.__result = self.__readString()
    self.__info = self.__readString()
    return self.__read() == 0

  # Returns the result.
  def result(self):
    return self.__result

  # Returns the info string.
  def info(self):
    return self.__info

  # Closes the socket.
  def close(self):
    s.send('exit \0')
    s.close()

  # Initializes the byte transfer
  def __init(self):
    self.__bpos = 0
    self.__bsize = 0

  # Receives a string from the socket.
  def __readString(self):
    bf = array.array('B')
    while True:
      b = self.__read()
      if b:
        bf.append(b)
      else:
        return bf.tostring()

  # Returns the next byte.
  def __read(self):
    # Cache next bytes
    if self.__bpos == self.__bsize:
      self.__bsize = s.recv_into(self.__buf)
      self.__bpos = 0

    # Return current byte
    b = self.__buf[self.__bpos]
    self.__bpos += 1
    return b


# This class offers an interactive BaseX console.
class Client(object):
  # Initializes the client.
  def __init__(self, host, port):
    self.__host = host
    self.__port = port
    print 'BaseX Client'
    print 'Try "help" to get some information.'

  # Creates a session.
  def session(self):
    user = raw_input('Username: ');
    pw = getpass.getpass('Password: ');
    try:
      global session
      session = Session(self.__host, self.__port, user, pw)
      self.__console()
      session.close()
      print 'See you.'
    except IOError as e:
      print e

  # Runs the console.
  def __console(self):
    session.execute('SET INFO ON')
    while True:
      com = str(raw_input('> ')).strip()
      if com == 'exit':
        break
      if com:
        session.execute(com)
        print session.info()

# Reads arguments -p and -h.
def opts():
    try:
      opts, args = getopt.getopt(sys.argv[1:], '-p:-h', ['port', 'host'])
    except getopt.GetoptError, err:
      print str(err)
      sys.exit()
    global host
    global port
    host = 'localhost'
    port = 1984

    for o, a in opts:
      if o == '-p':
        port = int(a)
      if o == '-h':
        host = a

# Main method.
if __name__ == '__main__':
  opts()
  Client(host, port).session()
